system/lib/wasmfs/wasmfs.cpp

```diff
@@ -0,0 +1,57 @@
+// // Copyright 2021 The Emscripten Authors.  All rights reserved.
+// // Emscripten is available under two separate licenses, the MIT license and the
+// // University of Illinois/NCSA Open Source License.  Both these licenses can be
+// // found in the LICENSE file.
+// // wasmfs.cpp will implement a new file system that replaces the existing JS filesystem.
+// // Current Status: Work in Progress.
+// // See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#include <stdlib.h>
+#define __NEED_struct_iovec
+#include <emscripten/emscripten.h>
+#include <wasi/api.h>
+
+extern "C" {
+
+__wasi_errno_t __wasi_fd_write(
+  __wasi_fd_t fd, const __wasi_ciovec_t* iovs, size_t iovs_len, __wasi_size_t* nwritten) {
+  // FD 1 = STDOUT and FD 2 = STDERR.
+  // Temporary hardcoding of filedescriptor values.
+  // TODO: May not want to proxy stderr (fd == 2) to the main thread.
+  // This will not not show in HTML, a console.warn in a worker is suffficient.
+  // This would be a change from the current FS.
+  if (fd == 1 || fd == 2) {
+    __wasi_size_t num = 0;
+    for (size_t i = 0; i < iovs_len; i++) {
+      const uint8_t* ptr = iovs[i].buf;
+      __wasi_size_t len = iovs[i].buf_len;
+      for (__wasi_size_t j = 0; j < len; j++) {
+        EM_ASM({ WASMFS.printChar($0, HEAPU8[$1]); }, fd, ptr + j);
```

Comment:
It would be better to send an entire buffer to JS, and not a single character at a time, as calls to JS are not cheap.

In fact this may be an issue if we get many tiny buffers. That is something we'll need to measure at some point I guess - maybe add a TODO.

Reply:
Changed it to just use the buffers that we receive via `iovs`